gadget/device.go

```diff
@@ -0,0 +1,73 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var ErrDeviceNotFound = errors.New("device not found")
+
+// FindDeviceForStructure attempts to find an existing device matching given
+// volume structure, by inspecting it's name and, optionally, the filesystem
+// label. Assumes that the host's udev has set up device symlinks correctly.
+func FindDeviceForStructure(ps *PositionedStructure) (string, error) {
+	var candidates []string
+
+	if ps.Name != "" {
+		byPartlabel := filepath.Join(dirs.GlobalRootDir, "/dev/disk/by-partlabel/", strutil.HexEscapePath(ps.Name))
+		candidates = append(candidates, byPartlabel)
+	}
+
+	if ps.Label != "" {
+		byFsLabel := filepath.Join(dirs.GlobalRootDir, "/dev/disk/by-label/", strutil.HexEscapePath(ps.Label))
+		candidates = append(candidates, byFsLabel)
+	}
+
+	var found string
+	for _, candidate := range candidates {
```

Comment:
Initially I thought that building and iterating over a two-element slice is strange, but it seems that doing it in a single pass would involve a lot of ugly conditionals and the result wouldn't be more readable anyway.